### ISSUE TITLE
dnnConversionCreate_F32 fails when running TF with optimized MKL

### ISSUE THREAD

**Issue by jakubkarczewski** [tensorflow/tensorflow#16982]
### System information
- **Have I written custom code**: Yes, (https://github.com/jakubkarczewski/AlexNetF/blob/master/alexnet.py)
- **OS Platform and Distribution**: Linux Centos 7
- **TensorFlow installed from**: compiled from source from https://github.com/tensorflow/tensorflow/releases
- **TensorFlow version**: 1.6.0-rc0
- **Python version**:  2.7
- **Bazel version**: 0.10.0
- **GCC/Compiler version**: stock Centos 7 gcc
- **Compilation command**: ```bazel build --config=mkl --copt="-DINTEL_MKL_ML" --copt="-mfma" --copt="-mavx2" --copt="-march=broadwell" --copt="-O3" -s -c opt //tensorflow/tools/pip_package:build_pip_package;;```
- **Exact command to reproduce**: python alexnet.py --training_epoch=1 --model_version=1 output/
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A

### Describe the problem
Running following training with Tensorflow compiled with command specified above results in error: ```2018-02-12 23:40:38.088756: F tensorflow/core/kernels/mkl_lrn_op.cc:595] Check failed: dnnConversionCreate_F32( &convert_input, static_cast<dnnLayout_t>(inimage_shape.GetCurLayout()), lt_internal_input) == E_SUCCESS (-1 vs. 0) ``` as opossed to running without any error and training properly on Tensorflow version available under ```pip install tensorflow```.
For training data I used Imagenet 60gb dataset (http://www.image-net.org/challenges/LSVRC/2012/) with 1000 classes.
What's more - following error can be found when running with Tensorflow from precompiled wheel files for both versions of python. This makes me think that the way I compile TF is not the problem here.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
CUDA/cuDNN version
GPU model and memory

**Comment by jakubkarczewski** [comment]
@tensorflowbutler of course, it should be ok now.

**Comment by cy89** [comment]
@jakubkarczewski so to clarify, is it the set of MKL-enabling flags that you think breaks your case? I.e., can you please give us the minimally contrasting build command that doesn't fail?

**Comment by jakubkarczewski** [comment]
Yes, I think the problem lies in those MKL flags. I've tried to compile Tensorflow in the same environment as mentioned above but with command ```bazel build --config=mkl --copt=”-DEIGEN_USE_VML” -c opt //tensorflow/tools/pip_package:
build_pip_package``` and got in return ``` ./tensorflow/core/util/mkl_util.h:710] Non-OK-status: Status(error::Code::UNIMPLEMENTED, "Unimplemented conversion function") status: Unimplemented: Unimplemented conversion function ```. It was a mistake on my side (haven't passed which MKL I actually wanted to use with ```--copt="-DINTEL_MKL_ML"``` ) but it proves the problem is MKL related. I think the version of TF available on pip is build with ```bazel build --config=opt //tensorflow/tools/pip_package:build_pip_package``` so what differs from my command is the MKL flag.

**Comment by cy89** [comment]
@jakubkarczewski, one quick thing to double-check: you can run "python alexnet.py --training_epoch=1 --model_version=1 output/" without problems on the non-MKL build, right?

**Comment by cy89** [comment]
A colleague suggests: The error appears to be E_INCORRECT_INPUT_PARAMETER.  It sounds to me like it might be a shape mismatch of some kind - so perhaps adding logging to see what the two layouts are?

(they traced through dnnConversionCreate_F32 to the Intel site which has https://software.intel.com/en-us/mkl-developer-reference-c-dnnconversioncreate, and then in turn found dnnError_t, which has the following cases:

typedef enum {
    E_SUCCESS                   =  0,
    E_INCORRECT_INPUT_PARAMETER = -1,
    E_UNEXPECTED_NULL_POINTER   = -2,
    E_MEMORY_ERROR              = -3,
    E_UNSUPPORTED_DIMENSION     = -4,
    E_UNIMPLEMENTED             = -127
} dnnError_t;

**Comment by jakubkarczewski** [comment]
@cy89 yes non-MKL build works just fine, I will follow your advice with adding more logging and will report results as soon as possible. Thanks for help!

**Comment by cy89** [comment]
Good luck!

**Comment by jakubkarczewski** [comment]
@cy89 I am having trouble with adding logging info because I can't seem to find the structure of ```dnnLayout_t```. Searching on github (https://github.com/search?p=3&q=_dnnLayout_s&type=Code&utf8=%E2%9C%93) I noticed it is assigned to empty structure which (**I guess**) means that the type of this variable is not strictly defined. Is there any way I can extract and print the layout's dimensions?

Maybe there is a way to do it from Python level? What I intended to do was to modify the  ```mkl_lrn_op.cc``` file so that it prints the dimensions of layouts to ```stderr``` and then compile the tensorflow with this modification. Although if it's possible to do from Python script, it would be great

**Comment by cy89** [comment]
@jakubkarczewski I'm sorry, I'm not an MKL expert. I agree that it looks like dnnLayout_t starts off as an empty structure.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by wei-v-wang** [comment]
Thank you @jakubkarczewski  for reporting and for going deeper into this debugging. We will reproduce and provide a fix for this. Please stay tuned.

**Comment by angersson** [comment]
@cy89 Can you keep an eye on the resolution of this issue?

**Comment by wei-v-wang** [comment]
@jakubkarczewski Could you please list the contents of the "/data/i1k-extracted/train" (10 lines of the output is fine). I am having difficulty make your code run with my imagenet dataset, so I wonder maybe you are using raw images?

**Comment by cy89** [comment]
@angersson happy to keep an eye out; I think we're mostly relying on Wei for help here.

**Comment by jakubkarczewski** [comment]
@wei-v-wang yes, I am using raw images. Directory `i1k-extracted/train/` contains directories named `0/, 1/, 2/, ..., 999/`. Each directory represents data for one particular class of images. In each one of them there are images in `.JPEG` format. Example name of 1st image from `0/` directory: `n02119789_10007.JPEG ` where `n02119789` stands for some kind of class id and `10007` is (I guess) image id.

The listing is as follows:
``` drwxrwxr-x 1002 jkarczew jkarczew 20480 Feb  9 09:52 ./
drwxrwxr-x    3 jkarczew jkarczew  4096 Feb  9 09:51 ../
drwxrwxr-x    2 jkarczew jkarczew  4096 Feb  9 09:52 0/
drwxrwxr-x    2 jkarczew jkarczew  4096 Feb  9 09:52 1/
drwxrwxr-x    2 jkarczew jkarczew  4096 Feb  9 09:52 10/
drwxrwxr-x    2 jkarczew jkarczew  4096 Feb  9 09:52 100/
drwxrwxr-x    2 jkarczew jkarczew  4096 Feb  9 09:52 101/
drwxrwxr-x    2 jkarczew jkarczew  4096 Feb  9 09:52 102/
drwxrwxr-x    2 jkarczew jkarczew  4096 Feb  9 09:52 103/
drwxrwxr-x    2 jkarczew jkarczew  4096 Feb  9 09:52 104/
```

**Comment by wei-v-wang** [comment]
Ah I see. Thank you for the info. Looks like maybe each year the imagenet are named differently -- e.g. my raw images had n02119789 being the directory names and yours are numbers like 0,1,10,... 
I will work on preparing identical dataset to yours. Thanks for the info!

**Comment by jakubkarczewski** [comment]
I might have used a preprocessed ImageNet dataset because I got it from a friend of mine. I bet he used mapping for subdirectories such as: https://gist.github.com/aaronpolhamus/964a4411c0906315deb9f4a3723aac57

**Comment by wei-v-wang** [comment]
@jakubkarczewski Please know that I was able to get the preprocessed ImageNet dataset identical to yours and was able to reproduce the error (e.g. use --config=mkl, by default MKL-DNN) as well as --config=mkl --copt="-DINTEL_MKL_ML" (MKL_ML). 
2018-04-05 23:28:14.468026: F ./tensorflow/core/util/mkl_util.h:710] Non-OK-status: Status(error::Code::UNIMPLEMENTED, "Unimplemented conversion function") status: Unimplemented: Unimplemented conversion function
It seems using MKL_ML our error message differs a little bit -- maybe it was caused by the commit difference. Hopefully fixing on MKL-DNN will also fix on the MKL_ML side. I will post the outcome now that I have reproduced the bug. Thanks for your patience!

**Comment by wei-v-wang** [comment]
As you originally reported: 
tensorflow/core/kernels/mkl_lrn_op.cc:595] Check failed: dnnConversionCreate_F32( &convert_input, static_cast<dnnLayout_t>(inimage_shape.GetCurLayout()), lt_internal_input) == E_SUCCESS (-1 vs. 0) 

The bug lies in LRN op as commenting out the two LRN layers in your script avoids the error. We will fix the LRN MKL ops. Please stay tuned. Thank you!

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by asimshankar** [comment]
(FYI @tatianashp )

**Comment by wei-v-wang** [comment]
@jakubkarczewski @cy89 @tatianashp @asimshankar 

Please see below for my newest findings regarding this issue - it seems the issue disappeared if we use Intel MKL DNN backend (with recent commits) rather than Intel MKL ML backend. 

AlexNet]$ export OMP_NUM_THREADS=4 (may need to do this to avoid OpenMP resource issue)
AlexNet]$ python alexnet.py   testoutput/
/usr/lib64/python2.7/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Loading data
Data loaded.
1281167
1000
WARNING:tensorflow:From alexnet.py:329: softmax_cross_entropy_with_logits (from tensorflow.python.ops.nn_ops) is deprecated and will be removed in a future version.
Instructions for updating:

Future major versions of TensorFlow will allow gradients to flow
into the labels input on backprop by default.

See tf.nn.softmax_cross_entropy_with_logits_v2.

Start time is: 1524900472.01
2018-04-28 00:27:52.015040: I tensorflow/core/platform/cpu_feature_guard.cc:140] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX512F
 Iter 0, Minibatch Loss= 6.898850, Training Accuracy= 0.0 Elapsed time:8.87320780754acc_up=[0.0, 0.0]
 Iter 20, Minibatch Loss= 6.916332, Training Accuracy= 0.0 Elapsed time:87.2397489548acc_up=[0.0, 0.0]

@jakubkarczewski Is it OK for you to use Intel MKL DNN? In order to use Intel MKL DNN backend, you
can just use the previous command you tried 
bazel build --config=mkl --copt=”-DEIGEN_USE_VML” -c opt //tensorflow/tools/pip_package: build_pip_package  (*please do not specify   --copt="-DINTEL_MKL_ML" * , this way by default will be INTEL_MKL_DNN). Sorry for the confusion. 

Could you please confirm?

**Comment by wei-v-wang** [comment]
@jakubkarczewski FYI, I used a commit that is very much like this public commit: https://github.com/tensorflow/tensorflow/commit/a175841eb549f069ac205fb32bf55314a387fe6d

**Comment by wei-v-wang** [comment]
@jakubkarczewski May I suggest trying v1.8.0 (https://github.com/tensorflow/tensorflow/commit/93bc2e2072e0daccbcff7a90d397b704a9e8f778) and  v170 tensorflow tagged version (https://github.com/tensorflow/tensorflow/releases/tag/v1.7.0)  and build with --config=mkl (by default it would be -DIntel_MKL_DNN, not -DIntel_MKLML) . I tried from my side, they do not have the issue as you reported anymore, please confirm. Thank you!

After you confirm, I suggest trying AlexNet in tf_cnn_benchmark (https://github.com/tensorflow/benchmarks/blob/master/scripts/tf_cnn_benchmarks/models/alexnet_model.py) instead - they run significantly faster than the keras one.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 17 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tatianashp** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
